pysecs/secs.py

```diff
@@ -120,31 +120,37 @@ def fit(self, obs_loc, obs_B, obs_var=None, epsilon=0.05):
         self.sec_amps_var = np.empty((ntimes, self.nsec))
 
         # Calculate the singular value decomposition (SVD)
-        # T_obs has shape (nobs, 3, nsec), we need to reshape it
-        # and add an extra dimension for time, making it (1, nobs*3, nsec)
-        # obs_var has shape (ntimes, nobs, 3), reshape it and add an
-        # extra dimension for the secs, making it (ntimes, nobs*3, 1)
-
-        # After all of the reshaping and division, the full svd_in matrix
-        # has shape (ntimes, nobs*3, nsec)
-        svd_in = (T_obs.reshape(-1, self.nsec)[np.newaxis, :] /
-                  obs_var.reshape((ntimes, -1))[..., np.newaxis])
+        # NOTE: T_obs has shape (nobs, 3, nsec), we reshape it
+        # to (nobs*3, nsec); obs_var has shape (ntimes, nobs, 3), 
+        # we reshape it to (ntimes, nobs*3), then loop over ntimes
+        # to solve using (potentially) time-dependent observation 
+        # error variances to weight the observations
         for i in range(ntimes):
-            # Iterate through all times, only calculating the SVD when needed
-            # The first timestep, or anytime the variance changes
-            if i == 0 or not np.all(svd_in[i, ...] == svd_in[i-1, ...]):
-                U, S, Vh = np.linalg.svd(svd_in[i, ...], full_matrices=False)
+            
+            # Only (re-)calculate SVD when necessary
+            if i == 0 or not np.all(obs_var[i] == obs_var[i-1]):
+
+                # Weight T_obs with obs_var
+                svd_in = (T_obs.reshape(-1, self.nsec) /
```

Comment:
Do you want to put this back into the `svd()` call itself since we don't use svd_in anywhere else?

Reply:
I considered this. I like breaking it out like this since the comment line makes it clear exactly what is happening. Also, as I understand things (there's a good chance I don't), whether we explicitly assign the weighted T_obs to a variable, or we rely on Python's implicit temporary variable when performing the operation in the call to svd(), makes no difference in terms of allocating memory, so there should be no difference in the processing overhead.

Reply:
👍  I'm just fine with that.

I agree with your assessment, there may be a very slight overhead by storing it as a Python variable, but not enough to make any difference.

The big change was not broadcasting to the entire time dimension, so this is excellent!